### PR TITLE
Initial amber microbenchmarks for performance testing core gpu features and functionality

### PR DESCRIPTION
**Pull request by petermcneeleychromium**
These benchmarks the performance of:
Intrinsic functions
Conditionals (branches)
Shared and SSBO memory
Cached memory access

crbug.com/361315331

### CODE REVIEW

File: tests/benchmarks/bandwidth/cache_random_access.amber [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+layout(set = 0, binding = 1) buffer BlockB {
+   uint data[];
+} 
+ssbo_write;
+
+
+uint ReadStrided(uint iter_val, uint ii){
+ // iter_val will always be zero 
+ const uint kStride =  1u;//14627u;// 3u;// 47u;//11u;//1u;
```
Comment (dj2) [inline tests/benchmarks/bandwidth/cache_random_access.amber:40]: Not sure what these comments mean?

File: tests/benchmarks/bandwidth/cache_random_access.amber [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
+
+layout(set = 0, binding = 0) buffer BlockA { 
+  uint data[];
+} ssbo_read;
+
+
+layout(set = 0, binding = 1) buffer BlockB {
+   uint data[];
+}
```
Comment (dj2) [inline tests/benchmarks/bandwidth/cache_random_access.amber:34]: nit: remove blank line

File: tests/benchmarks/bandwidth/cache_random_access.amber [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+   uint data[];
+} 
+ssbo_write;
+
+
+uint ReadStrided(uint iter_val, uint ii){
+ // iter_val will always be zero 
+ const uint kStride =  1u;//14627u;// 3u;// 47u;//11u;//1u;
+ const uint kSizeMask16M = 0xFFFFFF;
+ return ssbo_read.data[(iter_val + ii * kStride) & kSizeMask16M];
```
Comment (dj2) [inline tests/benchmarks/bandwidth/cache_random_access.amber:42]:
```suggestion
 return ssbo_read.data[(iter_val + (ii * kStride)) & kSizeMask16M];
```

File: tests/benchmarks/conditional/test_and_write.amber [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+} ssbo_init_data;
+
+layout(set = 0, binding = 1) buffer BlockB {
+  uint data[];
+} ssbo_write;
+
+uint TestAndOperate(uint iter_val, uint invx){
+  // We want this to be a real branch per thread
+  if(iter_val > invx){ 
+    iter_val = (iter_val*13)%6631 + 1024;
```
Comment (dj2) [inline tests/benchmarks/conditional/test_and_write.amber:43]:
```suggestion
    iter_val = ((iter_val*13)%6631) + 1024;
```

File: tests/benchmarks/conditional/test_and_write.amber [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    iter_val = (iter_val*213)%631 + 1024;
```
Comment (dj2) [inline tests/benchmarks/conditional/test_and_write.amber:44]:
```suggestion
    iter_val = ((iter_val*213)%631) + 1024;
```

**Comment by petermcneeleychromium** [comment]
I think we might need a solution to the amber formatting. I might just try a local VS plaintext formatter

File: tests/benchmarks/conditional/test_and_write.amber [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+layout(set = 0, binding = 1) buffer BlockB {
+  uint data[];
+} ssbo_write;
+
+uint TestAndOperate(uint iter_val, uint invx){
+  // We want this to be a real branch per thread
+  if(iter_val > invx){ 
+    // Simple LCGs that ensure iter_val is greater than 1024 
+    iter_val = ((iter_val*13)%6631) + 1024;
```
Comment (petermcneeleychromium) [inline tests/benchmarks/conditional/test_and_write.amber:44]: These constants need to be unsigned. I should perhaps make that clear.
I dont want negative modulus

File: tests/benchmarks/shared_local/internal_workgroup_ssbo.amber [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  uint data[];
+} ssbo_fake_volatile;
+
+
+void main() {
+    // This is required when using shared memory
+    if( gl_LocalInvocationID.x == 0){
+      ssbo_wr.data[0] = 0;
+    }
+    barrier();
```
Comment (petermcneeleychromium) [inline tests/benchmarks/shared_local/internal_workgroup_ssbo.amber:45]: Note i fixed this and made this only barrier

**Review by dj2 — APPROVED** [review]

**Pull request merged**